### ISSUE TITLE
feature request: redirects

### ISSUE THREAD

**Issue by Mithgol** [ipfs/notes#339]
This is a feature request that is not related to a specific implementation (such as go-ipfs) but proposes a feature for IPFS as a whole. Therefore it goes to the main repository (unless I've misunderstood something).

The proposed feature is the following: IPFS API (both CLI and REST) should have a way to store in IPFS (obtaining a hash, an IPFS URL, etc.) a resource which is not a file, but a redirect to the given URL.

When opening that resource by a web browser, any IPFS gate (such as a local gate at `localhost::8080`, or the default gateway `gateway.ipfs.io`, or any other) must redirect the browser to the given URL, using `302 Found` (or, optionally, `301 Moved Permanently`) as its HTTP response status code.

I believe that feature can have a couple of useful applications.

**Comment by MaxLaumeister** [comment]
I also want to see IPFS support 301 redirects.

The vast majority of the web is comprised of web pages that change from time to time, and while IPFS has immutability at its heart, IPNS is a layer that's proven very useful for mutable content.

I'd like IPFS to get to the point where I can host my website on it and have all the first-class features I currently get with my web host, including 301 and 302 redirects and custom 404/410 pages

From my understanding, HTTP status codes are not really integrated with IPFS since it's a separate protocol, but I think it would be useful to have a compatibility layer. The more compatible IPFS is with HTTP, the faster it will be integrated into the rest of the web, and the faster we'll start seeing native IPFS links in search engine results and from the rest of the older HTTP web.

Right now I could use CloudFlare's bridge to host my .com domain on IPFS, but I would have to use `meta refresh` redirects to redirect users from old links, and make do with CloudFlare's stock 404 page.

By allowing IPNS entries to redirect visitors in an unambiguous, 301-like way, IPFS gets to have its cake and eat it too. The choice of whether to link a mutable IPNS resource or an immutable IPFS resource becomes the choice of the end-user. This way, IPFS keeps data available while still allowing content curation.

**Comment by saurik** [comment]
I agree with @MaxLaumeister that this is essentially a blocking issue for a lot of "port existing web page to IPFS" attempts.
